[FEATURE] SNS 로그인 구현 #5

jiwon2724 opened the pull request:
## 참고사항 
- 디자인 시스템이 아직 안나와서 폰트 및 디자인 관련 속성 padding, margin등 값은 하드코딩으로 해놨는데 추후에 수정하겠습니다!
- 모듈의 삭제 및 이동으로 PR은 + / - 합산 300줄을 초과한점 양해부탁드립니다:sob::sob: 
- dev branch의 업데이트 내용을 마지막에 반영해서 커밋 수가 조금 많아졌습니다 죄송합니다:sob:😭
- 아직 정해지지 않았거나, api 연동이 나오지 않은 부분은 `TODO`로 표시해뒀습니다!

## 내용
1. SNS 로그인 구현
2. feature-mymodel, core-XX 모듈 삭제
3. core, feature 모듈 생성

### Done!
 - [x] 카카오 로그인 구현
 - [x] 구글 로그인 구현
 - [x] 로그인 UI 구현
 - [x] 필요한 resource 등록

### 관련 Issue
#5

jiwon2724 committed:
[BASE]#5 - gradle 라이브러리, 버전 카탈로그 데이터 추가

jiwon2724 committed:
[BASE]#5 - module 삭제 및 수정, 추가

jiwon2724 committed:
[FEAT]#5 - 로그인 UI 구현 완료
- 로그인 화면에 들어가는 기초작업 포함
- 디자인 시스템이 아직 안나와서 적용은 추후에 해야 할 것 같습니다.

jiwon2724 committed:
[BASE]#5 - data 모듈 gradle 수정

jiwon2724 committed:
[FEAT]#5 - SNS(구글, 카카오) 로그인 구현 완료

jiwon2724 committed:
[BASE]#5 - 미사용 플러그인 제거

jiwon2724 committed:
[FEAT]#5 - ktlint 적용 코드 정리

jiwon2724 committed:
[FEAT]#5 - ktlint 적용

jiwon2724 committed:
[BASE]#5 - ktlint 컨벤션 적용

jiwon2724 committed:
[BASE]#5 - ktlint 컨벤션 적용

myung6024 commented on app/src/main/java/com/peonlee/ui/SplashActivity.kt:
```diff
@@ -0,0 +1,17 @@
+import com.peonlee.login.LoginActivity
+
+class SplashActivity : AppCompatActivity() {
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_splash)
+
+        startActivity(Intent(this, LoginActivity::class.java))
+    }
```
splash screen을 이용해서 구현한 것일까요? google에서 권장하는 방법으로 구현하면 좋을 것 같아요~!

https://developer.android.com/develop/ui/views/launch/splash-screen

jiwon2724 replied:
아직 적용은 안 했는데 반영해서 올려놓겠습니다~~!!:raised_hands:

jiwon2724 replied:
스플래쉬 스크린 API 적용해서 반영했습니다! 디자인이 아직 안나와서 Theme은 기본 설정만 해놨습니다!

myung6024 commented on core/ui/src/main/java/com/peonlee/core/ui/BaseActivity.kt:
```diff
@@ -0,0 +1,12 @@
+import android.os.Bundle
+import androidx.appcompat.app.AppCompatActivity
+
+class BaseActivity : AppCompatActivity() {
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_base)
+    }
+}
```
이 BaseActivity와 BaseFragment는 어떤 용도일까요?
아직 구현이 안되어 있는 것 같습니다

jiwon2724 replied:
core모듈에서 전역적으로 사용되는 Base만 미리 정의해 뒀습니다! LoginActivity에 적용시켜서 올려놓겠습니다!!!

myung6024 commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    private fun googleLogin() {
+        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
+            .requestEmail()
+            .build()
+
+        val mGoogleSignInClient = GoogleSignIn.getClient(this, gso)
```
gso는 풀네임으로 mGoogleSignInClient에서 m을 빼는 것이 정한 컨벤션에 맞을 것 같아요~!

jiwon2724 replied:
죄송합니다😭 컨벤션 더 신경써서 작업하겠습니다! 반영해서 올려놓을게요!

myung6024 commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,117 @@
+        } catch (e: ApiException) {
+            // TODO : 실패로직 수행
+        }
+    }
```
google result의 handle이라는 것을 명시하거나 Task<GoogleSignInAccount>의 kotlin extension으로 만들어도 좋을 것 같네요

myung6024 commented on feature/login/src/main/java/com/peonlee/login/extensions/ContextExtension.kt:
```diff
@@ -0,0 +1,14 @@
+import com.kakao.sdk.auth.model.OAuthToken
+import com.kakao.sdk.user.UserApiClient
+
+fun Context.loginWithKakaoAccount(callback: (token: OAuthToken?, error: Throwable?) -> Unit) {
+    return UserApiClient.instance.loginWithKakaoAccount(this, callback = callback)
+}
+
+fun Context.isKakaoTalkLoginAvailable(): Boolean {
+    return UserApiClient.instance.isKakaoTalkLoginAvailable(this)
+}
```
nit: 요 extension은 LoginActivity에서만 사용될 것 같은데 LoginActivity 안에 private로 있어도 괜찮지 않을까요?

jiwon2724 replied:
생각해보니 login 모듈안에서만 extensions 패키지가 사용되네요.. 말씀대로 굳이 패키지를 만들지 않고
LoginActivity에서 private으로 사용되어도 상관 없을 것 같습니다!

myung6024 commented on feature/login/src/main/res/layout/activity_login.xml:
```diff
@@ -0,0 +1,62 @@
+        android:layout_width="0dp"
+        android:layout_height="56dp"
+        android:layout_marginStart="20dp"
+        android:layout_marginEnd="20dp"
```
start와 end가 같으면 marginHorizontal을 이용할 수 있을 것 같아요

myung6024 commented on feature/login/src/main/res/layout/activity_login.xml:
```diff
@@ -0,0 +1,62 @@
+        android:layout_marginBottom="14dp"
+        android:background="#123123"
+        android:gravity="center"
+        android:text="테스트 카카오 버튼"
```
테스트용 text는 tools를 이용하면 좋습니다!

myung6024 commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,117 @@
+        }
+    }
+
+    private fun init() {
+        googleSignInResult()
+
+        binding.tvGoogleLogin.setOnClickListener {
+            googleLogin()
+        }
+
+        binding.tvKakaoLogin.setOnClickListener {
+            kakaoLogin()
+        }
+    }
+}
```
nit : 이건 개인적인 의견이긴 한데 init이 처음에 사용되는데 코드는 가장 아래에 있어서 어떻게 구현되어 있는지 한눈에 보기 어려웠습니다.  사용된 바로 아래에 함수를 넣어보면 어떨까요?

jiwon2724 replied:
네 동의합니다...! 가독성 부분 더 신경써서 작업하겠습니다!!ㅎㅎ

myung6024 commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,117 @@
+
+class LoginActivity : AppCompatActivity() {
+    private lateinit var binding: ActivityLoginBinding
+    private lateinit var googleSignInLauncher: ActivityResultLauncher<Intent>
```
googleSignInLauncher는 불변 상태로 있는 것이 좋아보이는데 바로 선언하거나 by lazy를 이용해보면 어떨까요?

jiwon2724 replied:
넵!! 의견 감사합니다ㅎㅎ 반영하겠습니답!😀

jiwon2724 replied:
생명주기의 상태가 Resumed인 경우에는 googleSignInLauncher를 호출 못하더라고요 ㅠㅠ 그래서 by lazy는 사용이 불가능하고, 선언 시 초기화해줬습니다!

myung6024 commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,117 @@
+import com.peonlee.login.extensions.isKakaoTalkLoginAvailable
+import com.peonlee.login.extensions.loginWithKakaoAccount
+
+class LoginActivity : AppCompatActivity() {
```
아마 Token을 받아서 서버에 넘겨주는 과정이 이후에 필요할 것 같은데 미리 ViewModel 관련 작업을 해놓으면 어떨까요?

jiwon2724 replied:
넵!! 반영해서 올려놓겠습니다! 다음 feature때도 미리 작업해서 올릴게요~~!

myung6024 commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,117 @@
+            UserApiClient.instance.loginWithKakaoTalk(this) { token, error ->
+                when {
+                    error != null -> {
+                        val userLoginCancel = error is ClientError && error.reason == ClientErrorCause.Cancelled
```
userLoginCancel만 보면 어떤 형태일지 예상하기 어려운 것 같아요. 앞에 is를 prefix로 넣어주면 좋을 것 같아요

myung6024 commented on feature/login/src/main/res/layout/activity_login.xml:
```diff
@@ -0,0 +1,62 @@
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        app:layout_constraintVertical_bias="0.3"
+        app:srcCompat="@drawable/ic_launcher_background" />
```
src대신 srcCompat을 사용한 이유가 있을까요?

jiwon2724 replied:
아뇨 ㅠㅠ 잘못 설정했습니다! 바로 수정할게요!!!

myung6024 commented on feature/login/src/main/res/layout/activity_login.xml:
```diff
@@ -0,0 +1,62 @@
+        app:layout_constraintStart_toStartOf="parent" />
+
+    <ImageView
+        android:id="@+id/imageView"
```
위에는 tv_google_login 요기에서는 imageView로 id에서 사용하는 네이밍 컨벤션이 일치하지 않는 것 같아요!

jiwon2724 replied:
네!!! 따로 이벤트가 없어서 컨벤션 작업을 계속 놓치네요 ㅠㅠ 신경쓰겠습니다!!ㅎㅎ

myung6024 commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,117 @@
+        }
+    }
+
+    private fun init() {
```
`initViews() = with(binding) {`
이렇게 표현하면 아래에서는 binding을 생략해서 쓰기에 좋을 것 같아요

myung6024 commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,117 @@
+import android.app.Activity
+import android.content.Intent
+import android.os.Bundle
+import android.util.Log
```
요런 사용되지 않는 import들은 ktlintformat하면 알아서 지워줄거에요!

jiwon2724 replied:
ktlintformat을 사용하려구 했는데 사용시 원하지 않는 파일이 삭제될 수 있다고 해서 사용을 안했습니다!
사용해도 딱히 상관이 없을까요~?

aurora32s replied:
control + option + o 하시면 사용하지 않는 import 자동 제거됩니다. :-)

ckrudals commented on core/ui/build.gradle.kts:
```diff
@@ -0,0 +1,15 @@
+}
+
+dependencies {
+    implementation("androidx.appcompat:appcompat:1.6.1")
```
build-logic 에 정의되어 있는것을 활용하면 좋을 것 같아요.

aurora32s committed:
[EDIT] CI에 Kakao를 위한 설정 추가

aurora32s committed:
Merge branch 'dev' into feature/#5-sns-login

aurora32s committed:
[EDIT] app/MyApplication Ktlint 이슈 수정

28라인 들여쓰기 제거

jiwon2724 committed:
[REFACTOR]#5 - activity_login.xml 수정
- marginHorizontal 적용
- ImageView 컨벤션 적용
- srcCompat -> src 수정
- 테스트 TextView tools 적용

jiwon2724 committed:
[REFACTOR]#5 - LoginActivity.kt 수정
- ContextExtension.kt 제거, LoginActivity.kt에서 private으로 관리
- init() -> initViews() 이름, 위치 변경 with() 함수 적용
- userLoginCancel -> isUserLoginCancel 변수명 수정
- handleSignInResult() -> googleHandleSignInResult() 함수명 수정
- mGoogleSignInClient -> googleSignInClient 변수명 수정
- googleSignInLauncher read-only로 변경
- google_client_id key 숨김 완료

jiwon2724 committed:
[REFACTOR]#5 - BaseActivity.kt 구현 후 LoginActivity.kt에 적용
- BaseActivity 구현 및 LoginActivity 상속 후 적용

jiwon2724 committed:
[REFACTOR]#5 - 구글 로그인 코드 개선
- googleHandleSignInResult() 함수 삭제
- addOnCompleteListener 적용하여 성공, 실패 분기

jiwon2724 committed:
[REFACTOR]#5 - LoginViewModel 구현
- domain, data 약식 연동
- string.xml 문자열 리소스 정의 및 매핑

jiwon2724 committed:
[FEAT]#5 - SplashActivity.kt 구현
- SplashScreen API 적용

jiwon2724 committed:
Merge remote-tracking branch 'origin/feature/#5-sns-login' into feature/#5-sns-login

myung6024 commented on core/ui/src/main/java/com/peonlee/core/ui/BaseActivity.kt:
```diff
@@ -0,0 +1,25 @@
+        setContentView(binding.root)
+    }
+
+    abstract fun splashLoad()
```
splashLoad가 전역적으로 구현을 해야할까요?
어떤 용도로 사용되는 걸까요? 🤔

jiwon2724 replied:
일단 함수명 순서가 바뀐점 수정하겠습니다...!🙏🙏🙏

Splash테마를 app모듈의 manifest에서 적용하고 다른 모듈로 화면전환시 Manifest merger failed에러가 발생하더라구요 ㅠㅠ  

그래서 전역적으로 모듈별 화면마다 installSplashScreen()을 생각했었습니다!

그런데 생각해보니 manifest에서 activity 태그에 theme 속성을 정의해주면 될 것 같아서 테스트해봤는데, installSplashScreen() 함수 필요 없이 잘 작동되더라구요..ㅎㅎ

 splashLoad() 함수가 전역적으로 필요 없을 것 같은데, 삭제하고 manifest에서 theme을 등록해주는 방법으로 적용해도 괜찮을까요~?

myung6024 replied:
넵넵 진입하는 화면에만 installSplashScreen()이 있으면 되거든요. 다만 딥링크로 다른 화면에서도 진입이 가능한 경우에는 구현이 달라지기는 합니다.

jiwon2724 replied:
네 확인했습니다~~!😎 딥링크 관련 작업도 설정해야 하는 부분일까요~?.?

jiwon2724 replied:
해결완!

myung6024 commented on core/data/src/main/java/com/peonlee/data/login/LoginRepositoryImpl.kt:
```diff
@@ -0,0 +1,10 @@
+import com.peonlee.domain.login.LoginRepository
+import javax.inject.Inject
+
+class LoginRepositoryImpl @Inject constructor() : LoginRepository {
```
요거 네이밍이 Impl suffix와 앞에 Default, Fake 등을 prefix로 붙이는 네이밍 방식이 있는데 이 부분 이야기해보면 좋을 것 같아요

jiwon2724 replied:
네넵 확인했습니다!!👍👍

myung6024 commented:
요거 다른 커밋이랑 많이 겹쳐있어서 코드를 전체적으로 보기 어렵긴하네요. module 변경과 별도로 PR을 분리하면 좋았을 것 같아요.
우선 CI 기다렸다가 Approve 하겠습니다~!

jiwon2724 committed:
[REFACTOR]#5 - splash api 기능 개선

aurora32s commented on app/build.gradle.kts:
```diff
@@ -11,12 +12,15 @@ android {
+    implementation(project(":core:ui"))
+    implementation(project(":feature:login"))
+
+    implementation("androidx.appcompat:appcompat:1.6.1")
```
version catalog에서 관리할 수 있도록 하는게 좋을거 같아요.

aurora32s commented on app/google-services.json:
```diff
@@ -0,0 +1,47 @@
+{
```
google-services.json은 보안상 외부에 공개되지 않는 것이 좋을거 같아요.🤗
해당 google-services.json 문서를 보내주시면 CI에 적용해두도록 하겠습니다. :-)

jiwon2724 replied:
네 확인했습니다~ CI 복습겸 제가 올려놓을게요!:thumbsup:

aurora32s commented on app/src/main/java/com/peonlee/MyApplication.kt:
```diff
@@ -17,7 +17,14 @@
 
 @HiltAndroidApp
-class MyApplication : Application()
+class MyApplication : Application() {
```
Application 클래스의 이름을 MyApplication이 아닌, 프로젝트명이 포함되도록 짓는것은 어떠신가요?😌
( ex. PeonleeApplication )

jiwon2724 replied:
네 동의합니다! 템플릿의 디폴트 이름으로 들어간 것 같네요 ㅠ 반영해서 올릴게요!

aurora32s commented on settings.gradle.kts:
```diff
@@ -27,14 +27,17 @@ dependencyResolutionManagement {
-include(":feature-mymodel")
 include(":test-app")
+include(":feature:login")
+include(":core")
```
:core도 include할 필요가 있는걸까요?

jiwon2724 replied:
아뇨 불필요해 보입니다! 삭제 후 반영할게요!

aurora32s commented on feature/.gitignore:
```diff
@@ -0,0 +1 @@
+/build
```
feature 밑에는 모듈밖에 없어서 .gitignore은 없어도 될거 같은데 어떻게 생각하시나요?

jiwon2724 replied:
모듈 만들 때 디폴트로 들어가는 것 같은데, 필요 없을 것 같아요!

aurora32s commented on core/domain/build.gradle.kts:
```diff
@@ -0,0 +1,13 @@
+
+dependencies {
+    implementation(libs.javax.inject)
+}
```
관련된 부분은
- peonlee.android.library
- peonlee.android.hilt
에 정의되어 있어요.

그래서 이 부분에는 아래와 같이 plugin만 추가해주어도 동일하게 동작할 듯 합니다.
```
plugins {
    id("peonlee.android.library")
    id("peonlee.android.hilt")
}

android {
    namespace = "com.peonlee.core.domain"
}
```

aurora32s commented on gradle/libs.versions.toml:
```diff
@@ -13,32 +13,39 @@ androidxTestCore = "1.5.0"
+firebaseBom = "32.0.0"
+googleService = "4.3.15"
 hilt = "2.46"
+javaxInject = "1"
```
hilt 내에서 제공해주어서 별도로 추가할 필요 없을거 같아요. :-)

aurora32s commented on gradle/libs.versions.toml:
```diff
@@ -67,14 +74,26 @@ android-gradlePlugin = { group = "com.android.tools.build", name = "gradle", ver
+play-services-auth = { group = "com.google.android.gms", name = "play-services-auth", version.ref = "playServicesAuth"}
+
+# java
+javax-inject = { group = "javax.inject", name = "javax.inject", version.ref = "javaxInject"}
```
이 라이브러리도 hilt에서 제공해주므로 별도로 추가할 필요 없을듯 합니다.😌

aurora32s commented on core/domain/src/main/java/com/peonlee/domain/login/LoginRepository.kt:
```diff
@@ -0,0 +1,5 @@
+package com.peonlee.domain.login
+
+interface LoginRepository {
```
클린 아키텍처에 맞게 Repository interface는 domain module에 두신거 같아요. 이렇게 된다면 하위에 위치한 data module에서 반드시 domain module에 대한 의존성이 필요하게 되는 문제가 있을거 같아요.
Repository의 interface도 data module에 정의하는건 어떠신가요?
( * 참고로 nowinandroid에서도 Repository의 interfaces는 data module에 있습니다.🤗 )

jiwon2724 replied:
클린 아키텍처 기반으로 data module은 domain module에 대한 의존성이 필요한 것으로 알고있어요!
nowinandroid는 domain module에 data module이 들어가있는데 구글에서 권장하는 아키텍처가 맞을까요~?
https://developer.android.com/jetpack/guide?hl=ko

aurora32s replied:
엉클밥이 제안한 클린 아키텍처랑 Android 권장 아키텍처랑 약간 다르다고 알고 있기는 한데, 저는 이러한 아키텍처는 팀 내부에서 어느정도 맞추면 된다고 생각해서요.
Repository Interface를 Domain Module에 만들고, 그 구현 클래스를 Data Module에 만들게 되면 Domain Module에서는 Data Module에 대한 dependency 추가가 필요없게 되나요?

jiwon2724 replied:
네네 domain module은 순수 자바, 코틀린 언어만 들어가서 아무 종속성도 있으면 안되는 것으로 알고 있어요!
처음 기술 스택정할 때 클린 아키텍처 + 멀티 모듈이라고 정한 것 같아서 이 부분은 명준님이랑 같이 얘기 나눠봐요!!ㅎㅎ

aurora32s commented on core/domain/src/main/java/com/peonlee/domain/login/LoginUseCase.kt:
```diff
@@ -0,0 +1,9 @@
+
+import javax.inject.Inject
+
+class LoginUseCase @Inject constructor(private val loginRepository: LoginRepository) {
```
이건 그저 제 개인적인 취향입니다. :-)
constructor와 생성자 인자를 별도의 라인으로 관리하는건 어떠신가요?
👇🏻 요렇게!
```
class LoginUseCase @Inject constructor(
    private val loginRepository: LoginRepository
) {
    suspend fun login(token: String) {
        loginRepository.login(token)
    }
}
```

jiwon2724 replied:
LoginViewModel에선 들여쓰기가 잘 되어있네요 ㅎㅎㅠ 저도 동의합니다 더 신경쓸게요!

aurora32s commented on core/domain/src/main/java/com/peonlee/domain/login/LoginUseCase.kt:
```diff
@@ -0,0 +1,9 @@
+import javax.inject.Inject
+
+class LoginUseCase @Inject constructor(private val loginRepository: LoginRepository) {
+    suspend fun login(token: String) {
```
로그인 성공 / 실패와 관련된 로직은 추후에 추가하실 예정이신건가요?

jiwon2724 replied:
넵 맞습니답! 서버단의 우선순위가 크롤링이라 추후에 추가하려고 했어요!

jiwon2724 committed:
[Edit] - google_client_id 추가

google_client_id 추가

jiwon2724 committed:
[Edit] - google_client_id 수정

google_client_id 수정

jiwon2724 committed:
[Edit] - google_client_id 수정

google_client_id 수정

aurora32s commented on build.gradle.kts:
```diff
@@ -7,6 +13,7 @@ plugins {
     alias(libs.plugins.ksp) apply false
     alias(libs.plugins.kotlin.kapt) apply false
     alias(libs.plugins.ktlint.gradle) apply false
+    id("org.jetbrains.kotlin.jvm") version "1.8.20" apply false
```
이 plugin은 어떠한 이유로 추가하신건가요? 🤔

jiwon2724 replied:
테스트하다 추가로 넣은 것 같은데 필요 없을 것 같아요! 삭제하겠습니다~

jiwon2724 committed:
[Edit] - google_client_id 수정

jiwon2724 committed:
[REFACTOR]#5 - ktlint 컨벤션에 맞춰 코드 수정

aurora32s commented on feature/login/build.gradle.kts:
```diff
@@ -0,0 +1,32 @@
+    implementation(platform(libs.firebase.bom))
+    implementation(libs.firebase.analytics)
+    implementation(libs.play.services.auth)
+    implementation(libs.androidx.core.splashscreen)
```
feature module에서 splash screen 라이브러리가 필요하나요?

jiwon2724 replied:
splash screen api 작업시에 수정한 부분을 반영을 못했네요🙏 반영해서 다시 올리겠습니다~~!

aurora32s commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,106 @@
+
+@AndroidEntryPoint
+class LoginActivity : BaseActivity<ActivityLoginBinding>({ layoutInflater ->
+    ActivityLoginBinding.inflate(layoutInflater)
```
binding Factory를 BaseActivity로 넘기는 이유가 있을까요?

jiwon2724 replied:
viewBinding 생성 코드가 보일러 플레이트로 액티비티 개수마다 있을 것 같아서 선언해 주었어요! 
더 좋은 방법이 있을까요?.?

aurora32s replied:
요런 방법은 어떠신가요?🤗
```
override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    _binding = bindingFactory()
    setContentView(binding.root)
}

abstract fun bindingFactory(): T
```

aurora32s commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,106 @@
+                val task = GoogleSignIn.getSignedInAccountFromIntent(activityResult.data)
+                task.addOnCompleteListener { result ->
+                    if (result.isSuccessful) {
+                        val googleIdToken: String = task.result.idToken ?: ""
```
task의 idToken이 null 값일 때, 공백으로 viewModel에 넘기는게 맞나요? 🤔

jiwon2724 replied:
값이 공백인 경우 분기를 추가하겠습니다 감사합니다!😎

aurora32s commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,106 @@
+        }
+    }
+
+    private fun Context.loginWithKakaoAccount(callback: (token: OAuthToken?, error: Throwable?) -> Unit) {
```
LoginActivity 내에서만 호출되는 메서드인듯 한데, Context의 확장함수로 만든 이유가 있으실까요?

jiwon2724 replied:
`UserApiClient.instance` << 이 부분이 반복하여 사용되어서 정의해 봤어요! 너무 과하게 신경쓴 부분일까요?.?

aurora32s replied:
Activity도 결국 Context의 하위 클래스이므로 LoginActivity 내에서 실행된다면 일반함수로 만들어도 동일하게 동작할거 같아서요. :-)

jiwon2724 replied:
생각이 짧았습니다..! 확장함수를 굳이 사용안해도 될 것 같아요! 반영하겠습니다:thumbsup:

aurora32s commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    private fun Context.loginWithKakaoAccount(callback: (token: OAuthToken?, error: Throwable?) -> Unit) {
+        return UserApiClient.instance.loginWithKakaoAccount(this, callback = callback)
```
callback method를 kakaoLogin 메서드에서 만들고 해당 메서드의 인자로 전달하신 이유가 있을까요?

jiwon2724 replied:
카카오와 구글 로그인 두 개를 사용하는데, 카카오 로그인을 사용하지 않을 경우를 생각했어요! 
카카오 로그인을 사용할 경우에만 사용되게 만들었습니다!

aurora32s replied:
callback을 loginWithKakaoAccount 메서드 내에서
```
UserApiClient.instance.loginWithKakaoAccount(context=this) { token, error -> /* */ }
```
메서드 이름 자체가 KakaoAccount라서 카카오만을 위해 사용될거 같은데 요렇게 안하고 인수로 전달한 이유가 무엇인지 궁금합니다.

aurora32s commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,106 @@
+        return UserApiClient.instance.loginWithKakaoAccount(this, callback = callback)
+    }
+
+    private fun Context.isKakaoTalkLoginAvailable(): Boolean {
```
이 부분도 Context의 확장함수로 만든 이유가 궁금합니다. :-)

jiwon2724 replied:
위와 같은 이유에요! 이것도 수정하겠습니다~~!

aurora32s commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,106 @@
+        }
+
+        if (isKakaoTalkLoginAvailable()) {
+            UserApiClient.instance.loginWithKakaoTalk(this) { token, error ->
```
하단에 있는 Context.loginWithKakaoAccount 메서드 내의 UserApiClient.instance.loginWithKakaoAccount와 다른 기능으로 작동하는 건가요?
여기서 한번 loginWithKakaoAccount를 호출하고, callback에서 다시 loginWithKakaoAccount를 호출하는 이유가 무엇인가요?
( * 작동방법을 몰라 궁금해서 여쭤봐요 🤗 )

jiwon2724 replied:
아니요! 같은 기능으로 동작합니다!
카카오 로그인 문서에서는 카카오톡 설치여부에 따라서 카카오톡 로그인과 카카오 계정 로그인으로 나뉘고 있어요!

aurora32s replied:
제가 Kakao로그인 로직을 잘 몰라가지고😭
loginWithKakaoTalk 호출 후, error 발생 시 loginWithKakaoAccount를 호출하는건 카카오에서 권장하는 방법인건가요?

aurora32s commented on feature/login/src/main/java/com/peonlee/login/LoginActivity.kt:
```diff
@@ -0,0 +1,106 @@
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        initViews()
```
제 생각에는 initViews 호출이 모든 Activity에서 공통적으로 이루어질 거 같아서 BaseActivity에 있으면 어떨까 합니다!
👇🏻 BaseActivity에서
```
override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    _binding = bindingFactory(layoutInflater)
    setContentView(binding.root)
    initViews()
}

open fun initViews() { }
```
요렇게 해두면 LoginActivity에서는 initViews를 오버라이딩만 하면 되고,
onCreate는 정의할 필요가 없어질거 같아서, 제안드려보아요. 😆

aurora32s commented on core/ui/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,5 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android">
+    <application>
```
core module이여서 Manifest에 application은 없어도 될거 같아요. :-)

aurora32s commented on core/ui/src/main/res/layout/activity_base.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
```
(개인적인 궁금증)
activity_base.xml과 fragment_base.xml을 만들어두신 이유가 무엇인가요? 🤔

jiwon2724 replied:
`abstract` 붙이구 제거하는걸 까먹었습니다...ㅎ 삭제후 반영할게요!

aurora32s commented on feature/login/src/main/res/mipmap-anydpi-v26/ic_launcher.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<adaptive-icon xmlns:android="http://schemas.android.com/apk/res/android">
```
feature 모듈에 launcher icon을 추가해두신 이유가 있을까요?🤔

jiwon2724 replied:
feature모듈 생성시 화면이 있는 모듈로 만들어서 생긴 것 같아요 ㅠ 제거하겠습니다!

aurora32s commented on feature/login/src/main/res/values/colors.xml:
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <color name="black">#FF000000</color>
```
color와 같은 theme는 App 전체에서 공통적으로 관리하는게 좋을거 같아요. 
그래서 저는 보통 theme와 관련된 정보는 core.designsystem 또는 core.ui에 설정해두고 필요한 곳에서는 해당 module을 가져와 사용하는 방식으로 구현하고 있는데 어떠신가요?

jiwon2724 replied:
네 동의합니다~ 디자인 시스템 정해지고 세팅해놓을게요! feature모듈에 들어가는 모듈들의 colors등 디자인 관련된 파일은 제거할게요!

aurora32s commented on core/data/src/main/java/com/peonlee/data/login/LoginModule.kt:
```diff
@@ -0,0 +1,18 @@
+class LoginModule {
+    @Singleton
+    @Provides
+    fun provideLoginRepository(): LoginRepository {
```
@Provides보다 @Binds 로
```
@Binds
@Singleton
fun bindLoginRepository(
   loginRepositoryImpl: LoginRepositoryImpl
): LoginRepository
```
는 어떠신가요?

jiwon2724 replied:
혹시 @Binds를 더 선호하시는 이유가 있으실까요~?.?

aurora32s replied:
LoginRepository는 저희 프로젝트 내의 코드이기 때문에 생성자 주입으로 지정할 수 있기 때문입니다. :-)
@Provides는 Module 내에서 직접 객체를 생성하기 때문에 만약 LoginRepository에 의존성이 추가된다면 의존성을 직접 주입해야 하는 추가 작업이 생길 수도 있지 않을까 생각합니다.

aurora32s replied:
추가적으로 Hilt의 Module은 di package 하위로 위치시키는건 어떠신가요?

aurora32s commented on app/src/main/java/com/peonlee/ui/SplashActivity.kt:
```diff
@@ -0,0 +1,34 @@
+        content.viewTreeObserver.addOnPreDrawListener(
+            object : ViewTreeObserver.OnPreDrawListener {
+                override fun onPreDraw(): Boolean {
+                    return if (true) {
```
이 부분을 if - else 문으로 해둔 특별한 이유가 있을까요? 🤔

jiwon2724 replied:
아니요!
splash api 문서에 있는 코드를 가지고 왔어요! 디자인 확정나고 추후에 앱 방향에 맞게 바꾸면 될 것 같습니다 ㅎㅎ

aurora32s commented on build-logic/convention/src/main/kotlin/ext/configureSnsLogin.kt:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        dependencies {
+            "implementation"(libs.findLibrary("android.kakao.login").get())
```
해당 plugin이 sns 로그인을 위한 설정을 한 곳에서 관리하는 것이 목적이라면 여기에 feature.login의 build.gradle에 있는
```
implementation(libs.play.services.auth)
```
도 여기서 의존성 추가하는건 어떠신가요?🤗

aurora32s reviewed (commented):
지원님. sns 로그인 기능 구현하시느라 수고하셨습니다.👍🏻
코드를 보면서 궁금한 점이나 제안드리고 싶은 점에 대해 리뷰 남겨드렸어요.

추가적으로, 코드를 구현하실 때 어느 정도 주석을 추가해주시면 코드를 읽는 사람에게 도움이 되지 않을까 생각합니다. 🤗

감사합니다. 😌

jiwon2724 committed:
[REFACTOR]#5 - Application Class 이름 수정

jiwon2724 committed:
[REFACTOR]#5 - 미사용 resource 수정 및 제거

jiwon2724 committed:
[REFACTOR]#5 - 미사용 resource 수정 및 제거

jiwon2724 committed:
[REFACTOR]#5 - 카카오 로그인, BaseActivity 코드 개선
- manifestPlaceholders 수정
- manifest kakao activity 추가

jiwon2724 committed:
[REFACTOR]#5 - LoginActivity 주석 제거

jiwon2724 committed:
[REFACTOR]#5 - 버전 카탈로그, build.gradle.kts 수정

jiwon2724 committed:
[REFACTOR]#5 - repository 이동 (domain -> data)

jiwon2724 committed:
[REFACTOR]#5 - LoginModule 수정 (Provider -> Binds)

aurora32s reviewed (approved):
지원님. 요청내용 수정하시느라 수고하셨습니다. 🤗👍🏻
이전에 말씀드렸던 SplashActivity의 finish 처리 추가 부탁드립니다.

감사합니다. 😌

jiwon2724 committed:
[REFACTOR]#5 - SplashActivity finish() 추가
